pkg/apis/sources/v1alpha1/gcp_pubsub_types.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/knative/pkg/apis/duck"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Important: Run "make" to regenerate code after modifying this file
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ContainerSource is the Schema for the containersources API
+// +k8s:openapi-gen=true
+// +kubebuilder:categories=all,knative,eventing,sources
+type GcpPubSubSource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   GcpPubSubSourceSpec   `json:"spec,omitempty"`
+	Status GcpPubSubSourceStatus `json:"status,omitempty"`
+}
+
+// Check that ContainerSource can be validated and can be defaulted.
+var _ runtime.Object = (*GcpPubSubSource)(nil)
+
+// Check that ContainerSource implements the Conditions duck type.
+var _ = duck.VerifyType(&GcpPubSubSource{}, &duckv1alpha1.Conditions{})
+
+// ContainerSourceSpec defines the desired state of ContainerSource
+type GcpPubSubSourceSpec struct {
+	// GcpCredsSecret is the name of a K8s Secret in the same namespace as this GcpPubSubSource that
+	// contains the credentials used to authenticate with the GCP PubSub service. The credentials
+	// must be in the entry named 'key.json' and be in the JSON Service Account format.
+	GcpCredsSecret string `json:"gcpCredsSecret,omitempty`
```

Comment:
Could this be a subobject with name and key fields, like [`SecretKeySelector`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#secretkeyselector-v1-core)?

Reply:
Done.